platform/viewer/src/connectedComponents/ToolbarRow.js

```diff
@@ -150,46 +150,98 @@ class ToolbarRow extends Component {
+      !(Component.prototype && Component.prototype.isReactComponent)
+    );
+  }
 
-        return childButton;
-      });
+  function isClassComponent(Component) {
+    return !!(
+      typeof Component === 'function' &&
+      Component.prototype &&
+      Component.prototype.isReactComponent
+    );
+  }
+  // Check if its a valid customComponent. Later on an CustomToolbarComponent interface could be implemented.
+  if (
+    isClassComponent(CustomComponent) ||
+    isFunctionalComponent(CustomComponent)
+  ) {
+    const parentContext = this;
+    const isActive = activeButtons.includes(button.id);
 
-      return (
-        <ExpandableToolMenu
-          key={button.id}
-          label={button.label}
-          icon={button.icon}
-          buttons={childButtons}
-          activeCommand={activeCommand}
-        />
-      );
-    }
     return (
-      <ToolbarButton
+      <CustomComponent
+        parentContext={parentContext}
+        toolbarClickCallback={_handleToolbarButtonClick}
+        button={button}
         key={button.id}
-        label={button.label}
-        icon={button.icon}
-        onClick={_handleToolbarButtonClick.bind(this, button)}
-        isActive={activeButtons.includes(button.id)}
+        activeButtons={activeButtons}
+        isActive={isActive}
       />
     );
+  }
+}
+
+function _getExpandableButtonComponent(button, activeButtons) {
+  // Iterate over button definitions and update `onClick` behavior
+  let activeCommand;
+  const childButtons = button.buttons.map(childButton => {
+    childButton.onClick = _handleToolbarButtonClick.bind(this, childButton);
+
+    if (activeButtons.indexOf(childButton.id) > -1) {
+      activeCommand = childButton.id;
+    }
+
+    return childButton;
   });
+
+  return (
+    <ExpandableToolMenu
+      key={button.id}
+      label={button.label}
+      icon={button.icon}
+      buttons={childButtons}
+      activeCommand={activeCommand}
+    />
+  );
+}
+
+function _getDefaultButtonComponent(button, activeButtons) {
+  return (
+    <ToolbarButton
+      key={button.id}
+      label={button.label}
+      icon={button.icon}
+      onClick={_handleToolbarButtonClick.bind(this, button)}
+      isActive={activeButtons.includes(button.id)}
+    />
+  );
+}
+/**
+ * Determine which extension buttons should be showing, if they're
+ * active, and what their onClick behavior should be.
+ */
+function _getButtonComponents(toolbarButtons, activeButtons) {
+  const _this = this;
+  return toolbarButtons
+    .map(button => {
+      let getButtonComponentMethod = _getDefaultButtonComponent;
+
+      if (button.CustomComponent) {
+        getButtonComponentMethod = _getCustomButtonComponent;
+      } else if (button.buttons && button.buttons.length) {
+        getButtonComponentMethod = _getExpandableButtonComponent;
+      }
+
+      return getButtonComponentMethod.call(_this, button, activeButtons);
+    })
+    .filter(component => !!component);
```

Comment:
This is protecting against empty components? I think I'd rather it blew up so the user was aware of their configuration error.

Reply:
ok